### PR TITLE
Libarchive support

### PR DESCRIPTION
**Pull request by Vadiml1024**
Add support for archives supported by libarchive library

### CODE REVIEW

File: ratarmount.py [hunk 1]
```diff
@@ -18,7 +18,12 @@
 import urllib.parse
 import zipfile
 from typing import Any, Callable, Dict, Iterable, IO, List, Optional, Tuple, Union
-import fuse
+
+try:
```
Comment (mxmlnkn) [inline ratarmount.py:22]: Why use try-catch? It won't work anyway when fuse cannot be imported.

File: ratarmount.py [hunk 2]
```diff
@@ -549,6 +555,8 @@ def __init__(self, pathToMount: Union[str, List[str]], mountPoint: str, **option
                 self.selfBindMount = mountSource
                 self.mountPointFd = os.open(self.mountPoint, os.O_RDONLY)
 
+
```
Comment (mxmlnkn) [inline ratarmount.py:558]: Undo unrelated whitespace changes.

File: ratarmount.py [hunk 3]
```diff
@@ -1166,6 +1174,11 @@ def _parseArgs(rawArgs: Optional[List[str]] = None):
         '-d', '--debug', type=int, default=1,
         help='Sets the debugging level. Higher means more output. Currently, 3 is the highest.')
 
+    advancedGroup.add_argument(
+        '--force-libarchive', action='store_true', default=False,
+        help='Use libarchive to handle .tar, .rar, an .zip files.')
```
Comment (mxmlnkn) [inline ratarmount.py:1179]:
```suggestion
        help='Use libarchive to also handle TAR, RAR, and ZIP archives.')
```
I would say with leading dot refers to file extensions but ratarmount tries to be mostly extension agnostic, so call them by their names.

File: core/ratarmountcore/factory.py [hunk 4]
```diff
@@ -52,7 +79,7 @@ def openMountSource(fileOrPath: Union[str, IO[bytes]], **options) -> MountSource
     finally:
         if hasattr(fileOrPath, 'seek'):
             fileOrPath.seek(0)  # type: ignore
-
+
```
Comment (mxmlnkn) [inline core/ratarmountcore/factory.py:82]: Undo whitespace change (4 spaces were added)

File: core/ratarmountcore/compressions.py [hunk 5]
```diff
@@ -107,6 +114,14 @@
         lambda x: x.read(4) == (0xFD2FB528).to_bytes(4, 'little'),
         lambda x: indexed_zstd.IndexedZstdFile(x.fileno()),
     ),
+    '7z': CompressionInfo(
+        ['7z', '7zip'],
+        [],
+        'libarchive',
+        lambda x: x.read(6) == b'7z\xBC\xAF\x27\x1C',
+        lambda x: libarchive.SeekableArchive(x),
+        ),
```
Comment (mxmlnkn) [inline core/ratarmountcore/compressions.py:123]:
```suggestion
    ),
```

File: core/ratarmountcore/LibArchiveMountSource.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+
+import io
+import os
+import stat
+import time
+from typing import cast, Dict, IO, Iterable, List, Optional, Union
+
```
Comment (mxmlnkn) [inline core/ratarmountcore/LibArchiveMountSource.py:10]:
```suggestion

from .compressions import libarchive
```

I'm not 100% sure anymore why but I think this is to be preferred because else `libarchive` is undefined and maybe the style checker didn't like that.

Also, remove the try-import block on line 14-18 when using this.

File: core/ratarmountcore/LibArchiveMountSource.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+class FileInsideArchive(io.RawIOBase):
+    """
+    The seek implementation reopens the file on seeking back.
+    """
+
+    def __init__(self, reopen, file_size):
+        io.RawIOBase.__init__(self)
+        self.reopen = reopen
```
Comment (mxmlnkn) [inline core/ratarmountcore/LibArchiveMountSource.py:28]:
```suggestion
        self._reopen = reopen
```

Probably better to make it "private" as it currently isn't used from outside.

File: core/ratarmountcore/LibArchiveMountSource.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    def read(self, size: int = -1) -> bytes:
+        return self.fileobj.read(size)
+
+    def _skip(self, size: int) -> None:
+        BLKSIZE = 128 * 1024
+        while size > 0:
+            if size < BLKSIZE:
+                data = self.fileobj.read(size)
+            else:
+                data = self.fileobj.read(BLKSIZE)
```
Comment (mxmlnkn) [inline core/ratarmountcore/LibArchiveMountSource.py:69]:
```suggestion
            data = self.fileobj.read(min(size, BLKSIZE))
```

File: core/ratarmountcore/LibArchiveMountSource.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    # Basically copy paste of ZipMountSource because the interfaces are very similar
+
+    def __init__(self, fileOrPath: Union[str, IO[bytes]], **options) -> None:
+        self.fileObject : libarchive.SeekableArchive = libarchive.SeekableArchive(fileOrPath, mode='r')
+        LibArchiveMountSource._findPassword(self.fileObject, options.get("passwords", []))
+        self.files =  [e for e in self.fileObject]
+        self.options = options
+
+    @staticmethod
+    def _findPassword(fileobj: libarchive.SeekableArchive, passwords):
```
Comment (mxmlnkn) [inline core/ratarmountcore/LibArchiveMountSource.py:105]: Unused? Them, remove. Or, is it ToDo?

File: core/ratarmountcore/LibArchiveMountSource.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+    @overrides(MountSource)
+    def fileVersions(self, path: str) -> int:
+        return len(self._getFileInfos(path))
+
+    @overrides(MountSource)
+    def open(self, fileInfo: FileInfo) -> IO[bytes]:
+        e: libarchive.Entry = fileInfo.userdata[-1]
+        assert isinstance(e, libarchive.Entry)
+        return cast(IO[bytes], FileInsideArchive(lambda: self.fileObject.readstream(e.pathname), e.size))
```
Comment (mxmlnkn) [inline core/ratarmountcore/LibArchiveMountSource.py:199]:
```suggestion
        entry: libarchive.Entry = fileInfo.userdata[-1]
        assert isinstance(entry, libarchive.Entry)
        return cast(IO[bytes], FileInsideArchive(lambda: self.fileObject.readstream(entry.pathname),entry.size))
```

File: core/ratarmountcore/compressions.py [hunk 11]
```diff
@@ -107,6 +114,14 @@
         lambda x: x.read(4) == (0xFD2FB528).to_bytes(4, 'little'),
         lambda x: indexed_zstd.IndexedZstdFile(x.fileno()),
     ),
+    '7z': CompressionInfo(
+        ['7z', '7zip'],
```
Comment (mxmlnkn) [inline core/ratarmountcore/compressions.py:118]: This was bad design on my part but you should change the filter inside `SQLiteIndexedTar._detectCompression`. Instead of `if compressionId in ['rar', 'zip']:` it might be better to use a white list with: `if compressionId not in ['bz2', 'gz', 'xz', 'zst']:`. Without this there might be problems when trying to open a .7z file with SQLiteIndexedTar.

File: core/ratarmountcore/factory.py [hunk 12]
```diff
@@ -41,6 +42,32 @@ def openMountSource(fileOrPath: Union[str, IO[bytes]], **options) -> MountSource
                 [(lambda file=file: open(file, 'rb')) for file in filesToJoin]  # type: ignore
             )
 
+    forceLibarchive: bool = options.get("forceLibarchive", False)
+    special_formats = ('zip', 'tar', 'rar') if not forceLibarchive else (None, )
```
Comment (mxmlnkn) [inline core/ratarmountcore/factory.py:46]: What happens with xz, gzip, bz2, which are also seek-accelerated by SQLiteIndexedTar, and their tar variants (.tar.xz, ...)?

**Comment by mxmlnkn** [comment]
Please run the black code formatter. Personally, I'm running `tests/runtests.sh`, which also runs all code formatters and style checkers locally before running the actual tests.

File: ratarmount.py (hunk 1 above)
Comment (Vadiml1024) [reply ratarmount.py:22]: On my sytsem  import fuse is not recognized, i need to import fusepy

**Comment by mxmlnkn** [comment]
I've tried merging it, see the `libarchive` branch. But, there are too many conceptual open issues, some of them inside python-libarchive, and missing or failing tests. That's why it has not been merged into master yet nor develop yet. See also the extended todos posted in the chat group.
